crates/subspace-transaction-pool/src/bundle_validator.rs

```diff
@@ -0,0 +1,256 @@
+        let (retracted, enacted) = (route.retracted(), route.enacted());
+
+        // Remove bundles from the stale fork
+        for retracted_block in retracted {
+            match self.bundle_stored_in_last_k.front() {
+                Some((block_hash, _)) if *block_hash == retracted_block.hash => {
+                    self.bundle_stored_in_last_k.pop_front();
+                }
+                bb => {
+                    return Err(sp_blockchain::Error::Application(Box::from(
+                        format!(
+                            "Got wrong block from the bundle-collector, expect {:?}, got {:?}, this should not happen",
+                            retracted_block,
+                            bb.map(|(block_hash, _)| block_hash),
+                        ),
+                    )));
+                }
+            }
+        }
+
+        // Add bundles from the new block of the best fork
+        for enacted_block in enacted {
+            let block_bundles = self.collect_bundle_from_block(enacted_block.hash)?;
+            self.bundle_stored_in_last_k.push_front(block_bundles);
+        }
+
+        // Remove blocks from the back end to keep at most the bundle of the last K blocks
+        self.bundle_stored_in_last_k.truncate(self.confirm_depth_k);
+
+        Ok(())
+    }
+
+    pub async fn update_recent_stored_bundles(&mut self, new_best_hash: Block::Hash) {
+        let pre_best_hash = self.best_hash();
+        if let Err(err) = self.on_import_new_best_block(new_best_hash) {
+            tracing::error!(
+                %err,
+                "Failed to handle new best block for bundle-collector"
+            );
+        }
+        if pre_best_hash != self.best_hash() {
+            self.bundle_syncer
+                .update(self.bundle_stored_in_last_k.clone());
+        }
+    }
+}
+
+/// `BundleStoredInLastK` used to share the bundles of the last K blocks with different thread
+#[derive(Clone)]
+struct BundleStoredInLastK<Block: BlockT> {
+    #[allow(clippy::type_complexity)]
+    bundle_stored_in_last_k: Arc<Mutex<VecDeque<(Block::Hash, HashSet<Hash>)>>>,
+}
+
+impl<Block: BlockT> BundleStoredInLastK<Block> {
+    fn new() -> Self {
+        BundleStoredInLastK {
+            bundle_stored_in_last_k: Arc::new(Mutex::new(VecDeque::new())),
+        }
+    }
+
+    fn update(&self, incoming: VecDeque<(Block::Hash, HashSet<Hash>)>) {
+        *self.bundle_stored_in_last_k.lock() = incoming;
+    }
+}
+
+pub struct BundleValidator<Block: BlockT, Client> {
+    client: Arc<Client>,
+    bundle_syncer: BundleStoredInLastK<Block>,
+}
+
+impl<Block: BlockT, Client> Clone for BundleValidator<Block, Client> {
+    fn clone(&self) -> Self {
+        BundleValidator {
+            client: self.client.clone(),
+            bundle_syncer: self.bundle_syncer.clone(),
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct SkipBundleValidation;
```

Comment:
Cant we just use empty tuple and short circuit validation in default cases?

Reply:
An empty tuple was used initially, but it looks less straightforward on the call site, see https://github.com/subspace/subspace/blob/4033d3706b/domains/service/src/system_domain.rs#L187-L193, thus I suggested using something more explicit and descriptive. Feel free to suggest a better alternative.

Reply:
`()` is used originally, @liuchengxu suggested `SkipBundleValidation` as it is more informative I also agree with that.

Reply:
ahh I see. We always seem to use empty tuple for default impls. Explicit is fine as well

Reply:
Yeah, we use empty tuples a lot while constructing the runtime and feel comfortable there, but using `()` suddenly in such a place is somewhat not preferable.